packages/frontend/src/services/tokenExchange/RefFinanceContract.js

```diff
@@ -141,39 +145,76 @@ class RefFinanceContract {
                         ],
                     }),
                 },
-                contractConfig.gasLimit.swap,
+                refConfig.gasLimit.swap,
                 TOKEN_TRANSFER_DEPOSIT,
             ),
         );
 
         return actions;
     }
 
-    _formatPoolsData(inputPools) {
+    extractErrorMessage(data) {
+        if (!refConfig.errorRegExp) {
+            return '';
+        }
+
+        try {
+            return JSON.stringify(data).match(refConfig.errorRegExp);
+        } catch (error) {
+            console.error('Error on getting error message', error);
+        }
+
+        return '';
+    }
+
+    async _getTopPoolIds() {
+        if (!refConfig.indexerAddress) {
+            return;
+        }
+
+        if (localCache.topPoolIds) {
+            return localCache.topPoolIds;
+        }
+
+        const poolIds = new Set();
+
+        try {
+            const allPools = await fetch(
+                `${refConfig.indexerAddress}/list-top-pools`
+            ).then((res) => res.json());
+
+            allPools.forEach(({ id, tvl }) => {
+                if (Number(tvl) >= refConfig.pools.minimumTvlRequired) {
+                    poolIds.add(Number(id));
+                }
+            });
+            localCache.topPoolIds = poolIds;
+
+            return poolIds;
+        } catch (error) {
+            console.error('Error on fetching top pools', error);
+        }
+    }
+
+    async _formatPoolsData(inputPools) {
         const pools = {};
         const tokens = new Set();
+        const topPoolIds = await this._getTopPoolIds();
```

Comment:
underscore before the function name might be a sign that function is redundant inside exported class. Will be cool to move it away from class. Again not a blocker